// backtracker/dist/Frame.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
const path_1 = __importDefault(require("path"));
class Frame {
    path;
    dir;
    filename;
    srcPath;
    srcDir;
    srcFilename;
    srcLine;
    srcColumn;
    async;
    scope;
    line;
    column;
    anonymous;
    parent;
    constructor(details) {
        this.path = details.path;
        this.dir = path_1.default.dirname(details.path);
        this.filename = path_1.default.basename(details.path);
        this.srcPath = details.srcPath;
        this.srcDir = path_1.default.dirname(details.srcPath);
        this.srcFilename = path_1.default.basename(details.srcPath);
        this.srcLine = details.srcLine;
        this.srcColumn = details.srcColumn;
        this.async = details.async;
        this.scope = details.scope;
        this.line = details.line;
        this.column = details.column;
        this.anonymous = details.anonymous;
        this.parent = null;
    }
    toString() {
        return this.filename;
    }
}
module.exports = Frame;
